InteractionBase/src/vis/Menu.h

```diff
@@ -0,0 +1,119 @@
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+#pragma once
+
+#include "interactionbase_api.h"
+#include "ModelBase/src/nodes/Node.h"
+#include "VisualizationBase/src/declarative/DeclarativeItem.h"
+#include "MenuStyle.h"
+
+namespace Visualization {
+	class Text;
+}
+namespace Interaction {
+
+/**
+ * This class provides a standard interface for creating a menu which is opened
+ * at the last cursor position. Any kind of node can be used and actions may be
+ * associated with the nodes. The interface can also optionally contain a text field
+ * which is located above the normal items, where any text can be entered.
+ */
+class INTERACTIONBASE_API Menu : public Super<Visualization::DeclarativeItem<Menu>> {
+
+	ITEM_COMMON(Menu)
+
+	public:
+		virtual ~Menu();
+
+		/**
+		 * Whether this interface has a text field. Reimplement where necessary.
+		 */
+		virtual bool hasTextField() const;
+		Visualization::Text* textField() const;
+		/**
+		 * The currently focused item. Either the text field,
+		 * or a visualization of one of the nodes.
+		 */
+		Visualization::Item* focusedItem() const;
+		/**
+		 * Selects the given item as the focused item.
+		 */
+		void selectItem(Visualization::Item* item);
+		/**
+		 * Executes the currently focused item's function.
+		 * Returns whether it was successfully executed.
+		 */
+		bool executeFocused();
+
+		static void initializeForms();
+		virtual int determineForm();
+
+		virtual void determineChildren() override;
+		virtual void updateGeometry(int availableWidth, int availableHeight) override;
+
+	protected:
+		virtual bool sceneEventFilter(QGraphicsItem* watched, QEvent* event) override;
+		/**
+		 * The function to execute when selecting the given node.
+		 */
+		virtual bool onSelectNode(Model::Node* node) = 0;
+		/**
+		 * The function to execute when selecting the given text from the text field.
+		 */
+		virtual bool onSelectText(QString text) = 0;
```

Comment:
So, looking at the pure virtual methods here, I'm even more in favor of getting rid the the text field. Then we're left only with nodes - entries. Eventually we'll make it so that each entry knows how to execute itself, and then we won't need any pure virtual methods (I can work on this last bit). Then we ideally won't even need a dedicated derived class for the ViewSwitcher, provided we can easily configure the Menu class to do what we want.